### PR TITLE
Support log media configuration

### PR DESCRIPTION
**Pull request by buixor**
Add a `sp.log_media()` option, which supports "zend" and "syslog" as values, allowing the user to pick log media to be used by snuffleupagus.

The idea being that syslog won't/can't be intercepted by other php code such as in #307

TBD :
 - [x] Exit when SP_LOG_DROP
 - [x] Find a way to pass SP_LOG_DROP / SP_LOG_SIMULATION to syslog

### CODE REVIEW

File: src/sp_utils.c [hunk 1]
```diff
@@ -10,12 +10,23 @@ bool sp_zend_string_equals(const zend_string* s1, const zend_string* s2) {
...
-  zend_error(type, "[snuffleupagus][%s] %s", feature, msg);
+  switch (SNUFFLEUPAGUS_G(config).log_media) {
+  default:
+  case SP_ZEND:
+    zend_error(type, "[snuffleupagus][%s] %s", feature, msg);
+    break;
+  case SP_SYSLOG:
+    openlog("snuffleupagus", LOG_PID, LOG_AUTH);
+    error_filename = zend_get_executed_filename();
+    syslog(LOG_INFO, "[%s] %s in [%s]",feature, msg, error_filename);
```
Comment (jvoisin) [inline src/sp_utils.c]: Please add the line number too :)

**Comment by buixor** [comment]
@jvoisin : should we change syslog's message when SP_LOG_DROP, or relying on syslog level sounds good enough ?

**Comment by jvoisin** [comment]
It seems that this is the very purpose of [syslog levels]( https://en.wikipedia.org/wiki/Syslog#Severity_level ), so I think that it's better than good enough :)

File: src/sp_config_keywords.c [hunk 2]
```diff
@@ -78,11 +78,27 @@ int parse_session(char *line) {
...
 int parse_random(char *line) {
   return parse_enable(line, &(SNUFFLEUPAGUS_G(config).config_random->enable),
                       NULL);
 }
 
+int parse_log_media(char *line) {
+  size_t consumed = 0;
+  zend_string *value = get_param(&consumed, line, SP_TYPE_STR, SP_TOKEN_LOG_MEDIA);
+  
+  if (!strcmp(ZSTR_VAL(value), "zend")) {
```
Comment (jvoisin) [inline src/sp_config_keywords.c]: What about `php` instead of `zend`, since the later can be a bit confusing :)

File: src/sp_config_keywords.c [hunk 3]
```diff
@@ -78,11 +78,27 @@ int parse_session(char *line) {
...
 
+
 int parse_random(char *line) {
   return parse_enable(line, &(SNUFFLEUPAGUS_G(config).config_random->enable),
                       NULL);
 }
 
+int parse_log_media(char *line) {
+  size_t consumed = 0;
+  zend_string *value = get_param(&consumed, line, SP_TYPE_STR, SP_TOKEN_LOG_MEDIA);
```
Comment (jvoisin) [inline src/sp_config_keywords.c]: `get_param` can return `NULL`

File: src/sp_utils.c [hunk 4]
```diff
@@ -10,12 +10,27 @@ bool sp_zend_string_equals(const zend_string* s1, const zend_string* s2) {
...
-
-  zend_error(type, "[snuffleupagus][%s] %s", feature, msg);
+  switch (SNUFFLEUPAGUS_G(config).log_media) {
+  default:
+  case SP_ZEND:
+    zend_error(type, "[snuffleupagus][%s] %s", feature, msg);
+    break;
+  case SP_SYSLOG:
+    openlog("snuffleupagus", LOG_PID, LOG_AUTH);
+    error_filename = zend_get_executed_filename();
```
Comment (jvoisin) [inline src/sp_utils.c]: This variable could be declared here.

**Comment by kkadosh** [comment]
The code looks good to me.

I've updated the documentation for `sp.log_media`, @buixor @jvoisin can you please review it before merging please? :)

**Comment by jvoisin** [comment]
I think that we might want to log to syslog by default, since:

- php's log system can trivially be disabled once an attacker has arbitrary code execution
- it conflicts with some weird php behaviours, as outlined in #307

**Comment by buixor** [comment]
While I agree with you, I'm afraid it might be confusing for users :)

**Comment by jvoisin** [comment]
Feel free to add a test exercising the syslog logging mechanism before merging, otherwise, I'll likely add one afterwards anyway.

**Pull request merged**